src/execution/execute.ts

```diff
@@ -866,29 +906,62 @@ function ensureValidRuntimeType(
 
   if (typeof runtimeTypeName !== 'string') {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" must resolve to an Object type at runtime for field "${info.parentType.name}.${info.fieldName}" with ` +
-        `value ${inspect(result)}, received "${inspect(runtimeTypeName)}".`,
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" ` +
+        `with value ${inspect(result)} failed. ` +
+        `Encountered abstract type "${currentAbstractType.name}" must resolve to an Object or Interface type at runtime, ` +
+        `received "${inspect(runtimeTypeName)}".`,
+    );
+  }
+
+  if (encounteredTypeNames.has(runtimeTypeName)) {
+    throw new GraphQLError(
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
```

Comment:
We can't have cycles, they were possible only because previously you didn't do a subtype check (previous review comment).
Schema specifically validated to not have interface cycles, see schema validation code.
I already fixed this as part of previous comment will push on top of your PR.

Reply:
Pushed, please review.

Reply:
Some schemas are not validated i.e. "assumeValid" is true, and so I imagined we might still need the cycle check. I could be off-base.

Reply:
i added a test for an unvalidated schema to show why this is still necessary.

Reply:
As the comment says it's up to a person who sets the `assumeValid` flag is a guarantee that schema is valid.
https://github.com/graphql/graphql-js/blob/7532a5c8eb4001b726a2b45c8760bb71a6e10f2e/src/type/schema.ts#L416-L418

We can't assume something is broken when a person explicitly states it's valid.
Especially in performance-sensitive code.

Reply:
I am basing off how unvalidated fragment cycles are not allowed to infinite loop.